File: Data Structures/Python/Singly Linked List.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class SinglyLinkedList:
    def __init__(self):
        self.head = None

    def insert_beginning(self, value):
        new_node = Node(value)
        new_node.next = self.head
        self.head = new_node
        print("Element inserted at the beginning successfully.")

    def delete_beginning(self):
        if self.head is None:
            print("List is empty. Cannot delete from the beginning.")
            return

        temp = self.head
        self.head = self.head.next
        temp = None
        print("Element deleted from the beginning successfully.")

    def insert_end(self, value):
        new_node = Node(value)

        if self.head is None:
            self.head = new_node
            print("Element inserted at the end successfully.")
            return

        temp = self.head
        while temp.next is not None:
            temp = temp.next

        temp.next = new_node
        print("Element inserted at the end successfully.")

    def delete_end(self):
        if self.head is None:
            print("List is empty. Cannot delete from the end.")
            return

        if self.head.next is None:
            self.head = None
            print("Element deleted from the end successfully.")
            return

        temp = self.head
        prev = None
        while temp.next is not None:
            prev = temp
            temp = temp.next

        prev.next = None
        temp = None
        print("Element deleted from the end successfully.")

    def insert_middle(self, value):
        print("Enter the position to insert after: ")
        position = int(input())

        new_node = Node(value)

        temp = self.head
        for i in range(1, position):
            if temp is None:
                print("Invalid position. Element not inserted.")
                return
            temp = temp.next

        new_node.next = temp.next
        temp.next = new_node
        print(f"Element inserted at position {position} successfully.")

    def delete_middle(self):
        if self.head is None:
            print("List is empty. Cannot delete from the middle.")
            return

        print("Enter the position to delete: ")
        position = int(input())

        temp = self.head
        prev = None
        for i in range(1, position):
            if temp is None:
                print("Invalid position. Element not deleted.")
                return
            prev = temp
            temp = temp.next

        if prev is None:
            self.head = temp.next
        else:
            prev.next = temp.next

        temp = None
        print(f"Element deleted from position {position} successfully.")

    def display_list(self):
        temp = self.head
        print("List: ", end="")
        while temp is not None:
            print(temp.data, end=" ")
            temp = temp.next
        print()

if __name__ == "__main__":
    linked_list = SinglyLinkedList()

    while True:
        print("\n***** Singly Linked List Operations *****")
        print("1. Insert Beginning")
        print("2. Delete Beginning")
        print("3. Insert End")
        print("4. Delete End")
        print("5. Insert Middle")
        print("6. Delete Middle")
        print("7. Display List")
        print("8. Exit")

        choice = int(input("Enter your choice: "))

        if choice == 1:
            value = int(input("Enter the value to insert: "))
            linked_list.insert_beginning(value)
        elif choice == 2:
            linked_list.delete_beginning()
        elif choice == 3:
            value = int(input("Enter the value to insert: "))
            linked_list.insert_end(value)
        elif choice == 4:
            linked_list.delete_end()
        elif choice == 5:
            value = int(input("Enter the value to insert: "))
            linked_list.insert_middle(value)
        elif choice == 6:
            linked_list.delete_middle()
        elif choice == 7:
            linked_list.display_list()
        elif choice == 8:
            print("Exiting program. Goodbye!")
            break
        else:
            print("Invalid choice. Please enter a valid option.")

File: Data Structures/Java/Singly Linked List.java
import java.util.Scanner;

class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

public class SinglyLinkedList {
    private static Node head = null;

    public static void insertBeginning(int value) {
        Node newNode = new Node(value);
        newNode.next = head;
        head = newNode;
        System.out.println("Element inserted at the beginning successfully.");
    }

    public static void deleteBeginning() {
        if (head == null) {
            System.out.println("List is empty. Cannot delete from the beginning.");
            return;
        }

        Node temp = head;
        head = head.next;
        temp = null;
        System.out.println("Element deleted from the beginning successfully.");
    }

    public static void insertEnd(int value) {
        Node newNode = new Node(value);

        if (head == null) {
            head = newNode;
            System.out.println("Element inserted at the end successfully.");
            return;
        }

        Node temp = head;
        while (temp.next != null) {
            temp = temp.next;
        }

        temp.next = newNode;
        System.out.println("Element inserted at the end successfully.");
    }

    public static void deleteEnd() {
        if (head == null) {
            System.out.println("List is empty. Cannot delete from the end.");
            return;
        }

        if (head.next == null) {
            head = null;
            System.out.println("Element deleted from the end successfully.");
            return;
        }

        Node temp = head;
        Node prev = null;
        while (temp.next != null) {
            prev = temp;
            temp = temp.next;
        }

        prev.next = null;
        temp = null;
        System.out.println("Element deleted from the end successfully.");
    }

    public static void insertMiddle(int value) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the position to insert after: ");
        int position = scanner.nextInt();

        Node newNode = new Node(value);

        Node temp = head;
        for (int i = 1; i < position && temp != null; i++) {
            temp = temp.next;
        }

        if (temp == null) {
            System.out.println("Invalid position. Element not inserted.");
            return;
        }

        newNode.next = temp.next;
        temp.next = newNode;
        System.out.println("Element inserted at position " + position + " successfully.");
    }

    public static void deleteMiddle() {
        if (head == null) {
            System.out.println("List is empty. Cannot delete from the middle.");
            return;
        }

        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the position to delete: ");
        int position = scanner.nextInt();

        Node temp = head;
        Node prev = null;
        for (int i = 1; i < position && temp != null; i++) {
            prev = temp;
            temp = temp.next;
        }

        if (temp == null) {
            System.out.println("Invalid position. Element not deleted.");
            return;
        }

        if (prev == null) {
            head = temp.next;
        } else {
            prev.next = temp.next;
        }

        temp = null;
        System.out.println("Element deleted from position " + position + " successfully.");
    }

    public static void displayList() {
        Node temp = head;
        System.out.print("List: ");
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int choice, value;

        do {
            System.out.println("\n***** Singly Linked List Operations *****");
            System.out.println("1. Insert Beginning");
            System.out.println("2. Delete Beginning");
            System.out.println("3. Insert End");
            System.out.println("4. Delete End");
            System.out.println("5. Insert Middle");
            System.out.println("6. Delete Middle");
            System.out.println("7. Display List");
            System.out.println("8. Exit");

            System.out.print("Enter your choice: ");
            choice = scanner.nextInt();

            switch (choice) {
                case 1:
                    System.out.print("Enter the value to insert: ");
                    value = scanner.nextInt();
                    insertBeginning(value);
                    break;
                case 2:
                    deleteBeginning();
                    break;
                case 3:
                    System.out.print("Enter the value to insert: ");
                    value = scanner.nextInt();
                    insertEnd(value);
                    break;
                case 4:
                    deleteEnd();
                    break;
                case 5:
                    System.out.print("Enter the value to insert: ");
                    value = scanner.nextInt();
                    insertMiddle(value);
                    break;
                case 6:
                    deleteMiddle();
                    break;
                case 7:
                    displayList();
                    break;
                case 8:
                    System.out.println("Exiting program. Goodbye!");
                    break;
                default:
                    System.out.println("Invalid choice. Please enter a valid option.");
            }
        } while (choice != 8);
    }
}
